From cd68df08e2a7bc2116f9f32dc5918735d175edb4 Mon Sep 17 00:00:00 2001
From: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
Date: Wed, 26 Aug 2026 00:22:51 -0300
Subject: [PATCH] ARM: dts: airoha: Update EN7523 and EN7581 SoC device trees

Update memory reserved regions, add CPU OPP tables, configure thermal
zones, and include necessary peripheral definitions for Airoha EN7523
and EN7581 SoCs. Also, rename Ethernet reset names in MIPS device trees
to correct naming conventions.

Signed-off-by: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
---
 arch/arm/boot/dts/airoha/en7523-evb.dts   |   8 -
 arch/arm/boot/dts/airoha/en7523.dtsi      | 902 +++++++++++++++++++--
 arch/arm64/boot/dts/airoha/en7581-evb.dts |   4 +-
 arch/arm64/boot/dts/airoha/en7581.dtsi    | 910 ++++++++++++++++++++--
 arch/mips/boot/dts/econet/en751221.dtsi   |   2 +-
 arch/mips/boot/dts/econet/en7528.dtsi     |   2 +-
 6 files changed, 1673 insertions(+), 155 deletions(-)

diff --git a/arch/arm/boot/dts/airoha/en7523-evb.dts b/arch/arm/boot/dts/airoha/en7523-evb.dts
index f23a25cce119..0a79f5f6c311 100644
--- a/arch/arm/boot/dts/airoha/en7523-evb.dts
+++ b/arch/arm/boot/dts/airoha/en7523-evb.dts
@@ -26,14 +26,6 @@ memory@80000000 {
 	};
 };
 
-&gpio0 {
-	status = "okay";
-};
-
-&gpio1 {
-	status = "okay";
-};
-
 &pcie0 {
 	status = "okay";
 };
diff --git a/arch/arm/boot/dts/airoha/en7523.dtsi b/arch/arm/boot/dts/airoha/en7523.dtsi
index 5c1c8501fa6e..35611455fab3 100644
--- a/arch/arm/boot/dts/airoha/en7523.dtsi
+++ b/arch/arm/boot/dts/airoha/en7523.dtsi
@@ -2,9 +2,14 @@
 
 #include <dt-bindings/interrupt-controller/irq.h>
 #include <dt-bindings/interrupt-controller/arm-gic.h>
-#include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/clock/en7523-clk.h>
 #include <dt-bindings/reset/airoha,en7523-reset.h>
+#include <dt-bindings/leds/common.h>
+#include <dt-bindings/soc/airoha,scu-ssr.h>
+#include <dt-bindings/thermal/thermal.h>
+#include <dt-bindings/phy/phy.h>
+#include <dt-bindings/phy/airoha,an7581-usb-phy.h>
+#include <dt-bindings/soc/airoha,soc-variant.h>
 
 / {
 	interrupt-parent = <&gic>;
@@ -16,29 +21,134 @@ reserved-memory {
 		#size-cells = <1>;
 		ranges;
 
-		npu_binary@84000000 {
+		atf@80000000 {
 			no-map;
-			reg = <0x84000000 0xA00000>;
+			reg = <0x80000000 0x40000>;
 		};
 
-		npu_flag@84B0000 {
+		npu_binary: npu-binary@84000000 {
 			no-map;
-			reg = <0x84B00000 0x100000>;
+			reg = <0x84000000 0x00100000>;
 		};
 
-		npu_pkt@85000000 {
+		npu_pkt: npu-pkt@84d00000 {
 			no-map;
-			reg = <0x85000000 0x1A00000>;
+			reg = <0x84d00000 0x01600000>;
 		};
 
-		npu_phyaddr@86B00000 {
+		qdma0_buf: qdma0-buf@86400000 {
 			no-map;
-			reg = <0x86B00000 0x100000>;
+			reg = <0x86400000 0x02000000>;
 		};
 
-		npu_rxdesc@86D00000 {
+		qdma1_buf: qdma1-buf@88400000 {
 			no-map;
-			reg = <0x86D00000 0x100000>;
+			reg = <0x88400000 0x01000000>;
+		};
+	};
+
+	memory@80000000 {
+		device_type = "memory";
+		reg = <0x80000000 0x40000000>;
+	};
+
+	cpu_opp_table: opp-table {
+		compatible = "operating-points-v2";
+		opp-shared;
+
+		opp-500000000 {
+			opp-hz = /bits/ 64 <500000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-550000000 {
+			opp-hz = /bits/ 64 <550000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-600000000 {
+			opp-hz = /bits/ 64 <600000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-650000000 {
+			opp-hz = /bits/ 64 <650000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-7000000000 {
+			opp-hz = /bits/ 64 <700000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-7500000000 {
+			opp-hz = /bits/ 64 <750000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-8000000000 {
+			opp-hz = /bits/ 64 <800000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-8500000000 {
+			opp-hz = /bits/ 64 <850000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-9000000000 {
+			opp-hz = /bits/ 64 <900000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-9500000000 {
+			opp-hz = /bits/ 64 <950000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-1000000000 {
+			opp-hz = /bits/ 64 <1000000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_1G>;
+		};
+
+		opp-1050000000 {
+			opp-hz = /bits/ 64 <1050000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_12G>;
+		};
+
+		opp-1100000000 {
+			opp-hz = /bits/ 64 <1100000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_12G>;
+		};
+
+		opp-1150000000 {
+			opp-hz = /bits/ 64 <1150000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_12G>;
+		};
+
+		opp-1200000000 {
+			opp-hz = /bits/ 64 <1200000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_12G>;
+		};
+
+		opp-1250000000 {
+			opp-hz = /bits/ 64 <1250000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_13G>;
+		};
+
+		opp-1300000000 {
+			opp-hz = /bits/ 64 <1300000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_13G>;
+		};
+
+		opp-1350000000 {
+			opp-hz = /bits/ 64 <1350000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_14G>;
+		};
+
+		opp-1400000000 {
+			opp-hz = /bits/ 64 <1400000000>;
+			opp-supported-hw = <AIROHA_CPU_HW_14G>;
 		};
 	};
 
@@ -66,42 +176,31 @@ cpu0: cpu@0 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a53";
 			reg = <0x0>;
+			operating-points-v2 = <&cpu_opp_table>;
 			enable-method = "psci";
-			clock-frequency = <80000000>;
+			clocks = <&scuclk EN7523_CLK_CPU>;
+			clock-names = "cpu";
 			next-level-cache = <&L2_0>;
+			#cooling-cells = <2>;
 		};
 
 		cpu1: cpu@1 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a53";
 			reg = <0x1>;
+			operating-points-v2 = <&cpu_opp_table>;
 			enable-method = "psci";
-			clock-frequency = <80000000>;
+			clocks = <&scuclk EN7523_CLK_CPU>;
+			clock-names = "cpu";
 			next-level-cache = <&L2_0>;
+			#cooling-cells = <2>;
 		};
 
 		L2_0: l2-cache0 {
 			compatible = "cache";
-			cache-level = <2>;
-			cache-unified;
 		};
 	};
 
-	chip_scu: syscon@1fa20000 {
-		compatible = "airoha,chip-scu", "syscon";
-		reg = <0x1fa20000 0x400>;
-	};
-
-	scuclk: system-controller@1fb00000 {
-		compatible = "airoha,en7523-scu", "airoha,scuclk", "syscon";
-		reg = <0x1fb00000 0x1000>,
-		      <0x1efbe000 0x800>;
-		reg-names = "np-scu", "mcucfg";
-		airoha,chip-scu = <&chip_scu>;
-		#clock-cells = <1>;
-		#reset-cells = <1>;
-	};
-
 	gic: interrupt-controller@9000000 {
 		compatible = "arm,gic-v3";
 		interrupt-controller;
@@ -114,6 +213,18 @@ gic: interrupt-controller@9000000 {
 		      <0x09500000 0x2000>,
 		      <0x09600000 0x20000>;
 		interrupts = <GIC_PPI 9 IRQ_TYPE_LEVEL_LOW>;
+
+		its: gic-its@09020000 {
+			compatible = "arm,gic-v3-its";
+			msi-controller;
+			#msi-cell = <1>;
+			reg = <0x090200000 0x20000>;
+		};
+	};
+
+	pmu {
+		compatible = "arm,cortex-a15-pmu";
+		interrupts = <GIC_PPI 7 IRQ_TYPE_LEVEL_HIGH>;
 	};
 
 	timer {
@@ -125,6 +236,153 @@ timer {
 			     <GIC_PPI 10 IRQ_TYPE_LEVEL_LOW>;
 	};
 
+	chip_scu: syscon@1fa20000 {
+		compatible = "airoha,chip-scu", "syscon";
+		reg = <0x1fa20000 0x400>;
+	};
+
+	scuclk: system-controller@1fb00000 {
+		compatible = "airoha,en7523-scu", "airoha,scuclk", "syscon";
+		reg = <0x1fb00000 0x1000>,
+		      <0x1efbe000 0x800>;
+		reg-names = "np-scu", "mcucfg";
+		airoha,chip-scu = <&chip_scu>;
+		#clock-cells = <1>;
+		#reset-cells = <1>;
+	};
+
+	dramc: memory-controller@1fb24000 {
+		compatible = "airoha,en7523-dramc";
+		reg = <0x1fb24000 0x1000>,
+		      <0x1fb28000 0x1000>;
+		reg-names = "dramc", "ddrphy";
+		airoha,chip-scu = <&chip_scu>;
+	};
+
+	system-controller@1fbf0200 {
+		compatible = "syscon", "simple-mfd";
+		reg = <0x1fbf0200 0xc0>;
+
+		pinctrl: pinctrl {
+			compatible = "airoha,en7523-pinctrl";
+
+			interrupt-parent = <&gic>;
+			interrupts = <GIC_SPI 26 IRQ_TYPE_LEVEL_HIGH>;
+
+			airoha,chip-scu = <&chip_scu>;
+
+			gpio-controller;
+			#gpio-cells = <2>;
+
+			interrupt-controller;
+			#interrupt-cells = <2>;
+
+			gpio-ranges = <&pinctrl 0 12 28>;
+
+			pcie0_rst_pins: pcie0-rst-pins {
+				conf {
+					pins = "pcie_reset0";
+					drive-open-drain = <1>;
+				};
+			};
+
+			pcie1_rst_pins: pcie1-rst-pins {
+				conf {
+					pins = "pcie_reset1";
+					drive-open-drain = <1>;
+				};
+			};
+
+			uart2_pins: uart2 {
+				mux {
+					function = "uart";
+					groups = "uart2";
+				};
+			};
+
+			mdio_pins: mdio-pins {
+				mux {
+					function = "mdio";
+					groups = "mdio";
+				};
+
+				conf {
+					pins = "gpio9";
+					output-high;
+				};
+			};
+
+			pon_pinctrl: pon {
+				mux {
+					function = "pon";
+					groups = "pon";
+				};
+			};
+
+			swp1_led0: switch_port-1-led0 {
+				mux {
+					function = "phy1_led0";
+					pins = "gpio22";
+				};
+			};
+
+			swp1_led1: switch_port-1-led1 {
+				mux {
+					function = "phy1_led1";
+					pins = "gpio7";
+				};
+			};
+
+			swp2_led0: switch_port-2-led0 {
+				mux {
+					function = "phy2_led0";
+					pins = "gpio23";
+				};
+			};
+
+			swp2_led1: switch_port-2-led1 {
+				mux {
+					function = "phy2_led1";
+					pins = "gpio6";
+				};
+			};
+
+			swp3_led0: switch_port-3-led0 {
+				mux {
+					function = "phy3_led0";
+					pins = "gpio24";
+				};
+			};
+
+			swp3_led1: switch_port-3-led1 {
+				mux {
+					function = "phy3_led1";
+					pins = "gpio5";
+				};
+			};
+
+			swp4_led0: switch_port-4-led0 {
+				mux {
+					function = "phy4_led0";
+					pins = "gpio25";
+				};
+			};
+
+			swp4_led1: switch_port-4-led1 {
+				mux {
+					function = "phy4_led1";
+					pins = "gpio4";
+				};
+			};
+		};
+
+		pwm: pwm {
+			compatible = "airoha,en7581-pwm";
+
+			#pwm-cells = <3>;
+		};
+	};
+
 	uart1: serial@1fbf0000 {
 		compatible = "ns16550";
 		reg = <0x1fbf0000 0x30>;
@@ -132,27 +390,84 @@ uart1: serial@1fbf0000 {
 		reg-shift = <2>;
 		interrupts = <GIC_SPI 18 IRQ_TYPE_LEVEL_HIGH>;
 		clock-frequency = <1843200>;
-		status = "okay";
 	};
 
-	gpio0: gpio@1fbf0200 {
-		compatible = "airoha,en7523-gpio";
-		reg = <0x1fbf0204 0x4>,
-		      <0x1fbf0200 0x4>,
-		      <0x1fbf0220 0x4>,
-		      <0x1fbf0214 0x4>;
-		gpio-controller;
-		#gpio-cells = <2>;
+	uart2: serial@1fbf0300 {
+		compatible = "airoha,en7523-uart";
+		reg = <0x1fbf0300 0x30>;
+		interrupts = <GIC_SPI 32 IRQ_TYPE_LEVEL_HIGH>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&uart2_pins>;
+
+		status = "disabled";
+	};
+
+	watchdog@1fbf0100 {
+		compatible = "airoha,en7581-wdt";
+		reg = <0x1fbf0100 0x40>;
+
+		interrupts = <GIC_SPI 20 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 21 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 22 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
+
+		clocks = <&scuclk EN7523_CLK_BUS>;
+		clock-names = "bus";
+
+		#interrupt-cells = <1>;
 	};
 
-	gpio1: gpio@1fbf0270 {
-		compatible = "airoha,en7523-gpio";
-		reg = <0x1fbf0270 0x4>,
-		      <0x1fbf0260 0x4>,
-		      <0x1fbf0264 0x4>,
-		      <0x1fbf0278 0x4>;
-		gpio-controller;
-		#gpio-cells = <2>;
+	rng@1faa1000 {
+		compatible = "airoha,en7581-trng";
+		reg = <0x1faa1000 0xc04>;
+		interrupts = <GIC_SPI 35 IRQ_TYPE_LEVEL_HIGH>;
+	};
+
+	crypto@1e004000 {
+		compatible = "inside-secure,safexcel-eip93ies";
+		reg = <0x1fb70000 0x1000>;
+		interrupts = <GIC_SPI 44 IRQ_TYPE_LEVEL_HIGH>;
+	};
+
+	thermal: thermal-sensor@1efbd800 {
+		compatible = "airoha,en7523-thermal";
+		reg = <0x1efbd000 0x0fff>;
+		interrupts = <GIC_SPI 23 IRQ_TYPE_LEVEL_HIGH>;
+		airoha,chip-scu = <&chip_scu>;
+
+		#thermal-sensor-cells = <1>;
+	};
+
+	thermal-zones {
+		cpu_thermal: cpu-thermal {
+			polling-delay-passive = <250>;
+			polling-delay = <1000>;
+			thermal-sensors = <&thermal 0>;
+			trips {
+				cpu_passive: cpu-passive {
+					temperature = <300000>;
+					hysteresis = <0>;
+					type = "passive";
+				};
+			};
+			cooling-maps {
+				map0 {
+					trip = <&cpu_passive>;
+					cooling-device = <&cpu0 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu1 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+				};
+			};
+		};
+	};
+
+	pciephy: phy@1fa93700 {
+		compatible = "airoha,en7523-pcie-phy";
+		reg = <0x1fa93700 0x568>,
+		      <0x1fa95700 0x568>;
+		reg-names = "pcie0", "pcie1";
+		#phy-cells = <0>;
+		airoha,chip_scu = <&chip_scu>;
 	};
 
 	pcie0: pcie@1fa91000 {
@@ -168,9 +483,17 @@ pcie0: pcie@1fa91000 {
 		clocks = <&scuclk EN7523_CLK_PCIE>;
 		clock-names = "sys_ck0";
 		bus-range = <0x00 0xff>;
-		ranges = <0x82000000 0 0x20000000  0x20000000  0 0x8000000>;
+		ranges = <0x82000000 0 0x20000000 0x20000000 0 0x2000000>;
 		status = "disabled";
 
+		pinctrl-names = "default";
+		pinctrl-0 = <&pcie0_rst_pins>;
+
+		airoha,npu = <&npu>;
+
+		phys = <&pciephy>;
+		phy-names = "pcie-phy";
+
 		#interrupt-cells = <1>;
 		interrupt-map-mask = <0 0 0 7>;
 		interrupt-map = <0 0 0 1 &pcie_intc0 0>,
@@ -197,9 +520,17 @@ pcie1: pcie@1fa92000 {
 		clocks = <&scuclk EN7523_CLK_PCIE>;
 		clock-names = "sys_ck1";
 		bus-range = <0x00 0xff>;
-		ranges = <0x82000000 0 0x28000000  0x28000000  0 0x8000000>;
+		ranges = <0x82000000 0 0x22000000 0x22000000 0 0x2000000>;
 		status = "disabled";
 
+		pinctrl-names = "default";
+		pinctrl-0 = <&pcie1_rst_pins>;
+
+		airoha,npu = <&npu>;
+
+		phys = <&pciephy>;
+		phy-names = "pcie-phy";
+
 		#interrupt-cells = <1>;
 		interrupt-map-mask = <0 0 0 7>;
 		interrupt-map = <0 0 0 1 &pcie_intc1 0>,
@@ -213,23 +544,480 @@ pcie_intc1: interrupt-controller {
 		};
 	};
 
-	spi_ctrl: spi@1fa10000 {
-		compatible = "airoha,en7523-snand", "airoha,en7581-snand";
+	spi0: spi@1fa10000 {
+		compatible = "airoha,en7523-spi";
 		reg = <0x1fa10000 0x140>,
 		      <0x1fa11000 0x160>;
 
+		interrupts = <GIC_SPI 51 IRQ_TYPE_LEVEL_HIGH>;
 		clocks = <&scuclk EN7523_CLK_SPI>;
 		clock-names = "spi";
 
+		airoha,scu = <&scuclk>, <&chip_scu>;
+		airoha,scu-names = "scuclk", "chip-scu";
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+	};
+
+	usb0: usb@1fab0000 {
+		compatible = "mediatek,mtk-xhci";
+		reg = <0x1fab0000 0x3e00>,
+		      <0x1fab3e00 0x100>;
+		reg-names = "mac", "ippc";
+		interrupts = <GIC_SPI 33 IRQ_TYPE_LEVEL_HIGH>;
+
+		phys = <&usb0_phy PHY_TYPE_USB2>, <&usb0_phy PHY_TYPE_USB3>;
+
+		status = "disabled";
+	};
+
+	usb0_phy: phy@1fac0000 {
+		compatible = "airoha,an7581-usb-phy";
+		reg = <0x1fad0000 0x1fff>;
+
+		airoha,scu = <&scuclk>;
+		airoha,usb2-monitor-clk-sel = <AIROHA_USB2_MONCLK_SEL1>;
+		airoha,serdes-port = <AIROHA_SCU_SERDES_USB1>;
+
+		#phy-cells = <1>;
+	};
+
+	i2cclock: i2cclock@0 {
+		#clock-cells = <0>;
+		compatible = "fixed-clock";
+
+		/* 20 MHz */
+		clock-frequency = <20000000>;
+	};
+
+	i2c0: i2c0@1fbf8000 {
+		compatible = "airoha,en7523-i2c";
+		reg = <0x1fbf8000 0x100>;
+
+		clocks = <&i2cclock>;
+
+		/* 100 kHz */
+		clock-frequency = <100000>;
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		nand: nand@0 {
-			compatible = "spi-nand";
-			reg = <0>;
-			spi-max-frequency = <50000000>;
-			spi-tx-bus-width = <1>;
-			spi-rx-bus-width = <2>;
+		status = "disabled";
+	};
+
+	xpon_phy: phy@1faf0000 {
+		compatible = "airoha,en7523-xpon-phy";
+		reg = <0x1faf0000 0x5000>;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&pon_pinctrl>;
+
+		resets = <&scuclk EN7523_XPON_PHY_RST>;
+		reset-names = "phy";
+
+		airoha,scu = <&scuclk>;
+
+		#phy-cells = <0>;
+
+		status = "disabled";
+	};
+
+	xpon: xmac@1fb60000 {
+		compatible = "airoha,en7523-xpon";
+
+		reg = <0x1fb60000 0x10000>,
+		      <0x1fa2ff24 0x4>;
+		reg-names = "mac", "tx-off";
+
+		interrupts = <GIC_SPI 42 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 34 IRQ_TYPE_LEVEL_HIGH>;
+		interrupt-names = "mac", "dying-gasp";
+
+		resets = <&scuclk EN7523_XPON_MAC_RST>;
+		reset-names = "mac";
+
+		phys = <&xpon_phy>;
+		phy-names = "xpon";
+
+		ethernet = <&gdm2>;
+		airoha,scu = <&scuclk>;
+
+		status = "disabled";
+	};
+
+	pon_pcs: pcs@1fa08000 {
+		compatible = "airoha,en7523-pcs-pon";
+		reg = <0x1fa60000 0x1000>,
+		      <0x1fa65100 0x4a0>,
+		      <0x1fa65a00 0x1ac>,
+		      <0x1fa65e00 0x64>,
+		      <0x1fa66000 0xdc>;
+		reg-names = "pcs-mac", "pcs1", "pcs2",
+		            "an", "rate-adapt";
+
+		resets = <&scuclk EN7523_XPON_MAC_RST>,
+		         <&scuclk EN7523_XPON_PHY_RST>;
+		reset-names = "mac", "phy";
+
+		airoha,scu = <&scuclk>;
+
+		#pcs-cells = <0>;
+
+		status = "disabled";
+	};
+
+	pcie_pcs: pcs@1fa04000 {
+		compatible = "airoha,en7523-pcs-pcie";
+		reg = <0x1fa70000 0x1000>,
+		      <0x1fa71000 0x1000>,
+		      <0x1fa72100 0x4a0>,
+		      <0x1fa72a00 0x160>,
+		      <0x1fa72000 0x64>,
+		      <0x1fa72600 0xdc>,
+		      <0x1fa72c00 0x3b0>,
+		      <0x1fa77100 0x4a0>,
+		      <0x1fa77a00 0x160>,
+		      <0x1fa77000 0x64>,
+		      <0x1fa77600 0xdc>,
+		      <0x1fa77c00 0x3b0>;
+		reg-names = "pcs-mac0", "pcs-mac1",
+			    "pcie0-pcs1", "pcie0-pcs2", "pcie0-an",
+			    "pcie0-rate-adapt", "pcie0-phya",
+			    "pcie1-pcs1", "pcie1-pcs2", "pcie1-an",
+			    "pcie1-rate-adapt", "pcie1-phya";
+
+		airoha,scu = <&scuclk>;
+		#pcs-cells = <1>;
+		status = "disabled";
+	};
+
+	usb_pcs: pcs@1fa07000 {
+		compatible = "airoha,en7523-pcs-usb";
+		reg = <0x1fa80000 0x1000>,
+		      <0x1fa81100 0x4a0>,
+		      <0x1fa81a00 0x160>,
+		      <0x1fa81000 0x64>,
+		      <0x1fa81600 0xdc>,
+		      <0x1fa81c00 0x3b0>;
+		reg-names = "pcs-mac", "pcs1", "pcs2", "an",
+			    "rate-adapt", "phya";
+
+		phys = <&usb0_phy PHY_TYPE_USB3>;
+
+		airoha,scu = <&scuclk>;
+		#pcs-cells = <0>;
+		status = "disabled";
+	};
+
+	npu: npu@1e800000 {
+		compatible = "airoha,en7523-npu";
+		reg = <0x1e900000 0x313000>;
+		interrupts = <GIC_SPI 125 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 118 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 119 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 120 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 121 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 122 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 123 IRQ_TYPE_LEVEL_HIGH>;
+
+		airoha,scu = <&scuclk>;
+		memory-region = <&npu_binary>, <&npu_pkt>;
+		memory-region-names = "binary", "pkt";
+		/* Enable to debug NPU, remove on publish to upstream repository */
+		// airoha,enable_npu_tx_uart;
+		// airoha,npu-log;
+	};
+
+	wed0: wed@1fa02000 {
+		compatible = "airoha,en7523-wed", "syscon";
+		reg = <0x1fa02000 0xb00>,
+		      <0x1fa06000 0x400>;
+		reg-names = "wed", "wdma";
+		interrupts = <GIC_SPI 67 IRQ_TYPE_LEVEL_HIGH>;
+	};
+
+	wed1: wed@1fa03000 {
+		compatible = "airoha,en7523-wed", "syscon";
+		reg = <0x1fa03000 0xb00>,
+		      <0x1fa06400 0x400>;
+		reg-names = "wed", "wdma";
+		interrupts = <GIC_SPI 68 IRQ_TYPE_LEVEL_HIGH>;
+	};
+
+	eth: ethernet@1fb50000 {
+		compatible = "airoha,en7523-eth";
+		status = "disabled";
+		reg = <0x1fb50000 0x2600>,
+		      <0x1fb54000 0x2000>,
+		      <0x1fb56000 0x2000>,
+		      <0x1fbf9000 0x1000>;
+		reg-names = "fe", "qdma0", "qdma1", "gdmp";
+
+		resets = <&scuclk EN7523_FE_RST>,
+		         <&scuclk EN7523_FE_PDMA_RST>,
+		         <&scuclk EN7523_FE_QDMA_RST>,
+		         <&scuclk EN7523_DUAL_HSI0_MAC_RST>,
+		         <&scuclk EN7523_DUAL_HSI1_MAC_RST>,
+		         <&scuclk EN7523_HSI_MAC_RST>,
+		         <&scuclk EN7523_GDMP_RST>;
+		reset-names = "fe", "pdma", "qdma",
+		              "hsi0-mac", "hsi1-mac", "hsi-mac",
+									"gdmp";
+
+		interrupts = <GIC_SPI 37 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 55 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 56 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 57 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 38 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 58 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 59 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 60 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 49 IRQ_TYPE_LEVEL_HIGH>,
+		             <GIC_SPI 64 IRQ_TYPE_LEVEL_HIGH>;
+
+		memory-region = <&qdma0_buf>, <&qdma1_buf>;
+		memory-region-names = "qdma0-buf", "qdma1-buf";
+
+		airoha,npu = <&npu>;
+		airoha,wed = <&wed0>, <&wed1>;
+
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		gdm1: ethernet@1 {
+			compatible = "airoha,eth-mac";
+			reg = <1>;
+			phy-mode = "internal";
+			status = "disabled";
+
+			fixed-link {
+				speed = <10000>;
+				full-duplex;
+				pause;
+			};
+		};
+
+		gdm2: ethernet@2 {
+			compatible = "airoha,eth-mac";
+			reg = <2>;
+			pcs-handle = <&pon_pcs>;
+
+			status = "disabled";
+		};
+
+		mac3: ethernet@3 {
+			compatible = "airoha,eth-mac";
+			reg = <3>;
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			gdm3_pcie0: ethernet-port@4 {
+				compatible = "airoha,eth-port";
+				reg = <4>;
+				pcs-handle = <&pcie_pcs 0>;
+				status = "disabled";
+			};
+
+			gdm3_pcie1: ethernet-port@5 {
+				compatible = "airoha,eth-port";
+				reg = <5>;
+				pcs-handle = <&pcie_pcs 1>;
+				status = "disabled";
+			};
+
+			gdm3_usb: ethernet-port@6 {
+				compatible = "airoha,eth-port";
+				reg = <6>;
+				pcs-handle = <&usb_pcs>;
+				status = "disabled";
+			};
+		};
+	};
+
+	switch: switch@1fb58000 {
+		compatible = "airoha,en7523-switch";
+		reg = <0x1fb58000 0x8000>;
+		resets = <&scuclk EN7523_GSW_RST>;
+
+		interrupt-controller;
+		#interrupt-cells = <1>;
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_SPI 31 IRQ_TYPE_LEVEL_HIGH>;
+
+		status = "disabled";
+
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			gsw_port1: port@1 {
+				reg = <1>;
+				phy-mode = "internal";
+				phy-handle = <&gsw_phy1>;
+				status = "disabled";
+			};
+
+			gsw_port2: port@2 {
+				reg = <2>;
+				phy-mode = "internal";
+				phy-handle = <&gsw_phy2>;
+				status = "disabled";
+			};
+
+			gsw_port3: port@3 {
+				reg = <3>;
+				phy-mode = "internal";
+				phy-handle = <&gsw_phy3>;
+				status = "disabled";
+			};
+
+			gsw_port4: port@4 {
+				reg = <4>;
+				phy-mode = "internal";
+				phy-handle = <&gsw_phy4>;
+				status = "disabled";
+			};
+
+			gsw_port6: port@6 {
+				reg = <6>;
+				ethernet = <&gdm1>;
+				phy-mode = "internal";
+
+				fixed-link {
+					speed = <10000>;
+					full-duplex;
+					pause;
+				};
+			};
+		};
+
+		mdio: mdio {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			gsw_phy1: ethernet-phy@9 {
+				compatible = "ethernet-phy-ieee802.3-c22";
+				reg = <9>;
+				phy-mode = "internal";
+				status = "disabled";
+
+				pinctrl-names = "gbe-led";
+				pinctrl-0 = <&swp1_led0 &swp1_led1>;
+				airoha,scu = <&scuclk>;
+
+				leds {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					gsw_phy1_led0: led0@0 {
+						reg = <0>;
+						function = LED_FUNCTION_LAN;
+						status = "disabled";
+						active-low;
+					};
+
+					gsw_phy1_led1: led1@1 {
+						reg = <1>;
+						status = "disabled";
+						function = LED_FUNCTION_ACTIVITY;
+						active-low;
+					};
+				};
+			};
+
+			gsw_phy2: ethernet-phy@10 {
+				compatible = "ethernet-phy-ieee802.3-c22";
+				reg = <10>;
+				phy-mode = "internal";
+				status = "disabled";
+
+				pinctrl-names = "gbe-led";
+				pinctrl-0 = <&swp2_led0 &swp2_led1>;
+				airoha,scu = <&scuclk>;
+
+				leds {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					gsw_phy2_led0: led0@0 {
+						reg = <0>;
+						function = LED_FUNCTION_LAN;
+						status = "disabled";
+						active-low;
+					};
+
+					gsw_phy2_led1: led1@1 {
+						reg = <1>;
+						status = "disabled";
+						function = LED_FUNCTION_ACTIVITY;
+						active-low;
+					};
+				};
+			};
+
+			gsw_phy3: ethernet-phy@11 {
+				compatible = "ethernet-phy-ieee802.3-c22";
+				reg = <11>;
+				phy-mode = "internal";
+				status = "disabled";
+
+				pinctrl-names = "gbe-led";
+				pinctrl-0 = <&swp3_led0 &swp3_led1>;
+				airoha,scu = <&scuclk>;
+
+				leds {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					gsw_phy3_led0: led0@0 {
+						reg = <0>;
+						function = LED_FUNCTION_LAN;
+						status = "disabled";
+						active-low;
+					};
+
+					gsw_phy3_led1: led1@1 {
+						reg = <1>;
+						status = "disabled";
+						function = LED_FUNCTION_ACTIVITY;
+						active-low;
+					};
+				};
+			};
+
+			gsw_phy4: ethernet-phy@12 {
+				compatible = "ethernet-phy-ieee802.3-c22";
+				reg = <12>;
+				phy-mode = "internal";
+				status = "disabled";
+
+				pinctrl-names = "gbe-led";
+				pinctrl-0 = <&swp4_led0 &swp4_led1>;
+				airoha,scu = <&scuclk>;
+
+				leds {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					gsw_phy4_led0: led0@0 {
+						reg = <0>;
+						function = LED_FUNCTION_LAN;
+						status = "disabled";
+						active-low;
+					};
+
+					gsw_phy4_led1: led1@1 {
+						reg = <1>;
+						status = "disabled";
+						function = LED_FUNCTION_ACTIVITY;
+						active-low;
+					};
+				};
+			};
 		};
 	};
 };
diff --git a/arch/arm64/boot/dts/airoha/en7581-evb.dts b/arch/arm64/boot/dts/airoha/en7581-evb.dts
index dae9968a4ff6..4d0b2f262d31 100644
--- a/arch/arm64/boot/dts/airoha/en7581-evb.dts
+++ b/arch/arm64/boot/dts/airoha/en7581-evb.dts
@@ -65,9 +65,7 @@ reserved_bmt@7e00000 {
 	};
 };
 
-&en7581_pinctrl {
-	gpio-ranges = <&en7581_pinctrl 0 13 47>;
-
+&pinctrl {
 	pcie0_rst_pins: pcie0-rst-pins {
 		conf {
 			pins = "pcie_reset0";
diff --git a/arch/arm64/boot/dts/airoha/en7581.dtsi b/arch/arm64/boot/dts/airoha/en7581.dtsi
index ff6908a76e8e..30ddf2e3d006 100644
--- a/arch/arm64/boot/dts/airoha/en7581.dtsi
+++ b/arch/arm64/boot/dts/airoha/en7581.dtsi
@@ -3,7 +3,12 @@
 #include <dt-bindings/interrupt-controller/irq.h>
 #include <dt-bindings/interrupt-controller/arm-gic.h>
 #include <dt-bindings/clock/en7523-clk.h>
+#include <dt-bindings/phy/phy.h>
+#include <dt-bindings/phy/airoha,an7581-usb-phy.h>
+#include <dt-bindings/soc/airoha,scu-ssr.h>
 #include <dt-bindings/reset/airoha,en7581-reset.h>
+#include <dt-bindings/leds/common.h>
+#include <dt-bindings/thermal/thermal.h>
 
 / {
 	interrupt-parent = <&gic>;
@@ -15,29 +20,44 @@ reserved-memory {
 		#size-cells = <2>;
 		ranges;
 
-		npu-binary@84000000 {
+		atf: atf@80000000 {
+			no-map;
+			reg = <0x0 0x80000000 0x0 0x200000>;
+		};
+
+		npu_binary: npu-binary@84000000 {
 			no-map;
 			reg = <0x0 0x84000000 0x0 0xa00000>;
 		};
 
-		npu-flag@84b0000 {
+		qdma0_buf: qdma0-buf@87000000 {
+			no-map;
+			reg = <0x0 0x87000000 0x0 0x2000000>;
+		};
+
+		qdma1_buf: qdma1-buf@89000000 {
 			no-map;
-			reg = <0x0 0x84b00000 0x0 0x100000>;
+			reg = <0x0 0x89000000 0x0 0x1000000>;
 		};
 
-		npu-pkt@85000000 {
+		npu_pkt: npu-pkt@8a000000 {
 			no-map;
-			reg = <0x0 0x85000000 0x0 0x1a00000>;
+			reg = <0x0 0x8a000000 0x0 0x2c00000>;
 		};
 
-		npu-phyaddr@86b00000 {
+		npu_txpkt: npu-txpkt@8cc00000 {
 			no-map;
-			reg = <0x0 0x86b00000 0x0 0x100000>;
+			reg = <0x0 0x8cc00000 0x0 0x4000000>;
 		};
 
-		npu-rxdesc@86d00000 {
+		npu_txbufid: npu-txbufid@90c00000 {
 			no-map;
-			reg = <0x0 0x86d00000 0x0 0x100000>;
+			reg = <0x0 0x90c00000 0x0 0x6800>;
+		};
+
+		npu_ba: npu-ba@90c06800 {
+			no-map;
+			reg = <0x0 0x90c06800 0x0 0x200000>;
 		};
 	};
 
@@ -74,36 +94,56 @@ cpu0: cpu@0 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a53";
 			reg = <0x0>;
+			operating-points-v2 = <&cpu_opp_table>;
 			enable-method = "psci";
-			clock-frequency = <80000000>;
+			clocks = <&cpufreq>;
+			clock-names = "cpu";
+			power-domains = <&cpufreq>;
+			power-domain-names = "perf";
 			next-level-cache = <&l2>;
+			#cooling-cells = <2>;
 		};
 
 		cpu1: cpu@1 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a53";
 			reg = <0x1>;
+			operating-points-v2 = <&cpu_opp_table>;
 			enable-method = "psci";
-			clock-frequency = <80000000>;
+			clocks = <&cpufreq>;
+			clock-names = "cpu";
+			power-domains = <&cpufreq>;
+			power-domain-names = "perf";
 			next-level-cache = <&l2>;
+			#cooling-cells = <2>;
 		};
 
 		cpu2: cpu@2 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a53";
 			reg = <0x2>;
+			operating-points-v2 = <&cpu_opp_table>;
 			enable-method = "psci";
-			clock-frequency = <80000000>;
+			clocks = <&cpufreq>;
+			clock-names = "cpu";
+			power-domains = <&cpufreq>;
+			power-domain-names = "perf";
 			next-level-cache = <&l2>;
+			#cooling-cells = <2>;
 		};
 
 		cpu3: cpu@3 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a53";
 			reg = <0x3>;
+			operating-points-v2 = <&cpu_opp_table>;
 			enable-method = "psci";
-			clock-frequency = <80000000>;
+			clocks = <&cpufreq>;
+			clock-names = "cpu";
+			power-domains = <&cpufreq>;
+			power-domain-names = "perf";
 			next-level-cache = <&l2>;
+			#cooling-cells = <2>;
 		};
 
 		l2: l2-cache {
@@ -115,6 +155,159 @@ l2: l2-cache {
 		};
 	};
 
+	cpufreq: cpufreq {
+		compatible = "airoha,en7581-cpufreq";
+
+		operating-points-v2 = <&cpu_smcc_opp_table>;
+
+		#power-domain-cells = <0>;
+		#clock-cells = <0>;
+	};
+
+	cpu_opp_table: opp-table {
+		compatible = "operating-points-v2";
+		opp-shared;
+
+		opp-500000000 {
+			opp-hz = /bits/ 64 <500000000>;
+			required-opps = <&smcc_opp0>;
+		};
+
+		opp-550000000 {
+			opp-hz = /bits/ 64 <550000000>;
+			required-opps = <&smcc_opp1>;
+		};
+
+		opp-600000000 {
+			opp-hz = /bits/ 64 <600000000>;
+			required-opps = <&smcc_opp2>;
+		};
+
+		opp-650000000 {
+			opp-hz = /bits/ 64 <650000000>;
+			required-opps = <&smcc_opp3>;
+		};
+
+		opp-700000000 {
+			opp-hz = /bits/ 64 <700000000>;
+			required-opps = <&smcc_opp4>;
+		};
+
+		opp-750000000 {
+			opp-hz = /bits/ 64 <750000000>;
+			required-opps = <&smcc_opp5>;
+		};
+
+		opp-800000000 {
+			opp-hz = /bits/ 64 <800000000>;
+			required-opps = <&smcc_opp6>;
+		};
+
+		opp-850000000 {
+			opp-hz = /bits/ 64 <850000000>;
+			required-opps = <&smcc_opp7>;
+		};
+
+		opp-900000000 {
+			opp-hz = /bits/ 64 <900000000>;
+			required-opps = <&smcc_opp8>;
+		};
+
+		opp-950000000 {
+			opp-hz = /bits/ 64 <950000000>;
+			required-opps = <&smcc_opp9>;
+		};
+
+		opp-1000000000 {
+			opp-hz = /bits/ 64 <1000000000>;
+			required-opps = <&smcc_opp10>;
+		};
+
+		opp-1050000000 {
+			opp-hz = /bits/ 64 <1050000000>;
+			required-opps = <&smcc_opp11>;
+		};
+
+		opp-1100000000 {
+			opp-hz = /bits/ 64 <1100000000>;
+			required-opps = <&smcc_opp12>;
+		};
+
+		opp-1150000000 {
+			opp-hz = /bits/ 64 <1150000000>;
+			required-opps = <&smcc_opp13>;
+		};
+
+		opp-1200000000 {
+			opp-hz = /bits/ 64 <1200000000>;
+			required-opps = <&smcc_opp14>;
+		};
+	};
+
+	cpu_smcc_opp_table: opp-table-cpu-smcc {
+		compatible = "operating-points-v2";
+
+		smcc_opp0: opp0 {
+			opp-level = <0>;
+		};
+
+		smcc_opp1: opp1 {
+			opp-level = <1>;
+		};
+
+		smcc_opp2: opp2 {
+			opp-level = <2>;
+		};
+
+		smcc_opp3: opp3 {
+			opp-level = <3>;
+		};
+
+		smcc_opp4: opp4 {
+			opp-level = <4>;
+		};
+
+		smcc_opp5: opp5 {
+			opp-level = <5>;
+		};
+
+		smcc_opp6: opp6 {
+			opp-level = <6>;
+		};
+
+		smcc_opp7: opp7 {
+			opp-level = <7>;
+		};
+
+		smcc_opp8: opp8 {
+			opp-level = <8>;
+		};
+
+		smcc_opp9: opp9 {
+			opp-level = <9>;
+		};
+
+		smcc_opp10: opp10 {
+			opp-level = <10>;
+		};
+
+		smcc_opp11: opp11 {
+			opp-level = <11>;
+		};
+
+		smcc_opp12: opp12 {
+			opp-level = <12>;
+		};
+
+		smcc_opp13: opp13 {
+			opp-level = <13>;
+		};
+
+		smcc_opp14: opp14 {
+			opp-level = <14>;
+		};
+	};
+
 	timer {
 		compatible = "arm,armv8-timer";
 		interrupt-parent = <&gic>;
@@ -124,10 +317,62 @@ timer {
 			     <GIC_PPI 10 IRQ_TYPE_LEVEL_LOW>;
 	};
 
-	clk20m: clock-20000000 {
+	thermal-zones {
+		cpu_thermal: cpu-thermal {
+			polling-delay-passive = <0>;
+			polling-delay = <0>;
+
+			thermal-sensors = <&thermal 0>;
+
+			trips {
+				cpu_hot: cpu-hot {
+					temperature = <95000>;
+					hysteresis = <1000>;
+					type = "hot";
+				};
+
+				cpu-critical {
+					temperature = <120000>;
+					hysteresis = <1000>;
+					type = "critical";
+				};
+			};
+
+			cooling-maps {
+				map0 {
+					trip = <&cpu_hot>;
+					cooling-device = <&cpu0 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu1 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu2 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>,
+							 <&cpu3 THERMAL_NO_LIMIT THERMAL_NO_LIMIT>;
+				};
+			};
+		};
+	};
+
+	clk25m: oscillator {
 		compatible = "fixed-clock";
 		#clock-cells = <0>;
-		clock-frequency = <20000000>;
+		clock-frequency = <25000000>;
+		clock-output-names = "clkxtal";
+	};
+
+	vmmc_3v3: regulator-vmmc-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "vmmc";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-always-on;
+	};
+
+	sound: sound {
+		compatible = "airoha,an7581-wm8960-sound";
+
+		status = "disabled";
+
+		platform {
+			sound-dai = <&afe>;
+		};
 	};
 
 	soc {
@@ -150,7 +395,323 @@ gic: interrupt-controller@9000000 {
 			interrupts = <GIC_PPI 9 IRQ_TYPE_LEVEL_LOW>;
 		};
 
-		spi@1fa10000 {
+		uart1: serial@1fbf0000 {
+			compatible = "ns16550";
+			reg = <0x0 0x1fbf0000 0x0 0x30>;
+			reg-io-width = <4>;
+			reg-shift = <2>;
+			interrupts = <GIC_SPI 18 IRQ_TYPE_LEVEL_HIGH>;
+			clock-frequency = <1843200>;
+		};
+
+		watchdog@1fbf0100 {
+			compatible = "airoha,en7581-wdt";
+			reg = <0x0 0x1fbf0100 0x0 0x38>;
+
+			clocks = <&scuclk EN7523_CLK_BUS>;
+			clock-names = "bus";
+		};
+
+		uart2: serial@1fbf0300 {
+			compatible = "airoha,en7523-uart";
+			reg = <0x0 0x1fbf0300 0x0 0x30>;
+			reg-io-width = <4>;
+			reg-shift = <2>;
+			interrupts = <GIC_SPI 32 IRQ_TYPE_LEVEL_HIGH>;
+			clock-frequency = <7372800>;
+
+			status = "disabled";
+		};
+
+		hsuart3: serial@1fbe1000 {
+			compatible = "airoha,en7523-uart";
+			reg = <0x0 0x1fbe1000 0x0 0x40>;
+			reg-io-width = <4>;
+			reg-shift = <2>;
+			interrupts = <GIC_SPI 54 IRQ_TYPE_LEVEL_HIGH>;
+			clock-frequency = <7372800>;
+
+			status = "disabled";
+		};
+
+		afe: afe@1fbe2200 {
+			compatible = "airoha,an7581-afe";
+			reg = <0x0 0x1fbe2200 0x0 0x9000>;
+
+			interrupts = <GIC_SPI 48 IRQ_TYPE_LEVEL_HIGH>;
+			status = "disabled";
+		};
+
+		uart4: serial@1fbf0600 {
+			compatible = "airoha,en7523-uart";
+			reg = <0x0 0x1fbf0600 0x0 0x30>;
+			reg-io-width = <4>;
+			reg-shift = <2>;
+			interrupts = <GIC_SPI 61 IRQ_TYPE_LEVEL_HIGH>;
+			clock-frequency = <7372800>;
+
+			status = "disabled";
+		};
+
+		uart5: serial@1fbf0700 {
+			compatible = "airoha,en7523-uart";
+			reg = <0x0 0x1fbf0700 0x0 0x30>;
+			reg-io-width = <4>;
+			reg-shift = <2>;
+			interrupts = <GIC_SPI 18 IRQ_TYPE_LEVEL_HIGH>;
+			clock-frequency = <7372800>;
+
+			status = "disabled";
+		};
+
+		pon_pcs: pcs@1fa08000 {
+			compatible = "airoha,an7581-pcs-pon";
+			reg = <0x0 0x1fa08000 0x0 0x1000>,
+			      <0x0 0x1fa80000 0x0 0x60>,
+			      <0x0 0x1fa80a00 0x0 0x164>,
+			      <0x0 0x1fa84000 0x0 0x450>,
+			      <0x0 0x1fa85900 0x0 0x338>,
+			      <0x0 0x1fa86000 0x0 0x300>,
+			      <0x0 0x1fa8a000 0x0 0x1000>,
+			      <0x0 0x1fa8b000 0x0 0x1000>;
+			reg-names = "pcs_mac", "hsgmii_an", "hsgmii_pcs",
+				    "multi_sgmii", "usxgmii",
+				    "hsgmii_rate_adp", "pcs_ana", "pcs_pma";
+
+			resets = <&scuclk EN7581_XPON_MAC_RST>,
+				 <&scuclk EN7581_XPON_PHY_RST>;
+			reset-names = "mac", "phy";
+
+			airoha,scu = <&scuclk>;
+
+			#pcs-cells = <0>;
+		};
+
+		eth_pcs: pcs@1fa09000 {
+			compatible = "airoha,an7581-pcs-eth";
+			reg = <0x0 0x1fa09000 0x0 0x1000>,
+			      <0x0 0x1fa70000 0x0 0x60>,
+			      <0x0 0x1fa70a00 0x0 0x164>,
+			      <0x0 0x1fa74000 0x0 0x450>,
+			      <0x0 0x1fa75900 0x0 0x338>,
+			      <0x0 0x1fa76000 0x0 0x300>,
+			      <0x0 0x1fa7a000 0x0 0x1000>,
+			      <0x0 0x1fa7b000 0x0 0x1000>;
+			reg-names = "pcs_mac", "hsgmii_an", "hsgmii_pcs",
+				    "multi_sgmii", "usxgmii",
+				    "hsgmii_rate_adp", "pcs_ana", "pcs_pma";
+
+			resets = <&scuclk EN7581_XSI_MAC_RST>,
+				 <&scuclk EN7581_XSI_PHY_RST>;
+			reset-names = "mac", "phy";
+
+			airoha,scu = <&scuclk>;
+
+			#pcs-cells = <0>;
+		};
+
+		pcie_pcs: pcs@1fa04000 {
+			compatible = "airoha,an7581-pcs-pcie";
+			reg = <0x0 0x1fa04000 0x0 0x1000>,
+			      <0x0 0x1fa50000 0x0 0x60>,
+			      <0x0 0x1fa50a00 0x0 0x164>,
+			      <0x0 0x1fa54000 0x0 0x450>,
+			      <0x0 0x1fa55900 0x0 0x338>,
+			      <0x0 0x1fa56000 0x0 0x300>,
+			      <0x0 0x1fa05000 0x0 0x1000>,
+			      <0x0 0x1fa60000 0x0 0x60>,
+			      <0x0 0x1fa60a00 0x0 0x164>,
+			      <0x0 0x1fa64000 0x0 0x450>,
+			      <0x0 0x1fa65900 0x0 0x338>,
+			      <0x0 0x1fa66000 0x0 0x300>,
+			      <0x0 0x1fa5a000 0x0 0x1000>,
+			      <0x0 0x1fa5b000 0x0 0x1000>,
+			      <0x0 0x1fa5c000 0x0 0x1000>;
+			reg-names = "pcs_mac0", "hsgmii_an0", "hsgmii_pcs0",
+				    "multi_sgmii0", "usxgmii0",
+				    "hsgmii_rate_adp0",
+				    "pcs_mac1", "hsgmii_an1", "hsgmii_pcs1",
+				    "multi_sgmii1", "usxgmii1",
+				    "hsgmii_rate_adp1",
+				    "pcs_ana", "pcs_pma0", "pcs_pma1";
+
+			airoha,scu = <&scuclk>;
+
+			#pcs-cells = <1>;
+
+			status = "disabled";
+		};
+
+		usb_pcs: pcs@1fa07000 {
+			compatible = "airoha,an7581-pcs-usb";
+			reg = <0x0 0x1fa07000 0x0 0x1000>,
+			      <0x0 0x1fa90000 0x0 0x60>,
+			      <0x0 0x1fa90a00 0x0 0x164>,
+			      <0x0 0x1fa94000 0x0 0x450>,
+			      <0x0 0x1fa96000 0x0 0x300>,
+			      <0x0 0x1fa9a000 0x0 0x600>;
+			reg-names = "pcs_mac", "hsgmii_an", "hsgmii_pcs",
+				    "multi_sgmii",
+				    "hsgmii_rate_adp","pcs_ana";
+
+			phys = <&usb0_phy PHY_TYPE_USB3>;
+
+			airoha,scu = <&scuclk>;
+
+			#pcs-cells = <0>;
+
+			status = "disabled";
+		};
+
+		chip_scu: syscon@1fa20000 {
+			compatible = "airoha,en7581-chip-scu", "syscon";
+			reg = <0x0 0x1fa20000 0x0 0x388>;
+		};
+
+		pbus_csr: syscon@1fbe3400 {
+			compatible = "airoha,en7581-pbus-csr", "syscon";
+			reg = <0x0 0x1fbe3400 0x0 0xff>;
+		};
+
+		scuclk: clock-controller@1fb00000 {
+			compatible = "airoha,en7581-scu", "syscon";
+			reg = <0x0 0x1fb00000 0x0 0x970>;
+			#clock-cells = <1>;
+			#reset-cells = <1>;
+		};
+
+		rng@1faa1000 {
+			compatible = "airoha,en7581-trng";
+			reg = <0x0 0x1faa1000 0x0 0xc04>;
+			interrupts = <GIC_SPI 35 IRQ_TYPE_LEVEL_HIGH>;
+		};
+
+		usb0: usb@1fab0000 {
+			compatible = "mediatek,mtk-xhci";
+			reg = <0x0 0x1fab0000 0x0 0x3e00>,
+				<0x0 0x1fab3e00 0x0 0x100>;
+			reg-names = "mac", "ippc";
+			interrupts = <GIC_SPI 33 IRQ_TYPE_LEVEL_HIGH>;
+
+			phys = <&usb0_phy PHY_TYPE_USB2>, <&usb0_phy PHY_TYPE_USB3>;
+
+			status = "disabled";
+		};
+
+		usb0_phy: phy@1fac0000 {
+			compatible = "airoha,an7581-usb-phy";
+			reg = <0x0 0x1fac0000 0x0 0x10000>;
+
+			airoha,scu = <&scuclk>;
+			airoha,usb2-monitor-clk-sel = <AIROHA_USB2_MONCLK_SEL1>;
+			airoha,serdes-port = <AIROHA_SCU_SERDES_USB1>;
+
+			#phy-cells = <1>;
+		};
+
+		usb1: usb@1fad0000 {
+			compatible = "mediatek,mtk-xhci";
+			reg = <0x0 0x1fad0000 0x0 0x3e00>,
+				<0x0 0x1fad3e00 0x0 0x100>;
+			reg-names = "mac", "ippc";
+			interrupts = <GIC_SPI 150 IRQ_TYPE_LEVEL_HIGH>;
+
+			phys = <&usb1_phy PHY_TYPE_USB2>, <&usb1_phy PHY_TYPE_USB3>;
+
+			status = "disabled";
+		};
+
+		usb1_phy: phy@1fae0000 {
+			compatible = "airoha,an7581-usb-phy";
+			reg = <0x0 0x1fae0000 0x0 0x10000>;
+
+			airoha,scu = <&scuclk>;
+			airoha,usb2-monitor-clk-sel = <AIROHA_USB2_MONCLK_SEL2>;
+			airoha,serdes-port = <AIROHA_SCU_SERDES_USB2>;
+
+			#phy-cells = <1>;
+		};
+
+		crypto@1fb70000 {
+			compatible = "inside-secure,safexcel-eip93ies";
+			reg = <0x0 0x1fb70000 0x0 0x1000>;
+
+			interrupts = <GIC_SPI 44 IRQ_TYPE_LEVEL_HIGH>;
+		};
+
+		thermal: thermal-sensor@1efbd000 {
+			compatible = "airoha,en7581-thermal";
+			reg = <0x0 0x1efbd000 0x0 0xd5c>;
+			interrupts = <GIC_SPI 23 IRQ_TYPE_LEVEL_HIGH>;
+			airoha,chip-scu = <&chip_scu>;
+
+			#thermal-sensor-cells = <0>;
+		};
+
+		system-controller@1fbf0200 {
+			compatible = "syscon", "simple-mfd";
+			reg = <0x0 0x1fbf0200 0x0 0xc0>;
+
+			pinctrl: pinctrl {
+				compatible = "airoha,an7581-pinctrl";
+
+				interrupt-parent = <&gic>;
+				interrupts = <GIC_SPI 26 IRQ_TYPE_LEVEL_HIGH>;
+				airoha,chip-scu = <&chip_scu>;
+
+				gpio-ranges = <&pinctrl 0 13 47>;
+
+				gpio-controller;
+				#gpio-cells = <2>;
+
+				interrupt-controller;
+				#interrupt-cells = <2>;
+			};
+
+			en7581_pwm: pwm {
+				compatible = "airoha,en7581-pwm";
+
+				#pwm-cells = <3>;
+			};
+		};
+
+		i2cclock: i2cclock@0 {
+			#clock-cells = <0>;
+			compatible = "fixed-clock";
+
+			/* 20 MHz */
+			clock-frequency = <20000000>;
+		};
+
+		i2c0: i2c0@1fbf8000 {
+			compatible = "mediatek,mt7621-i2c";
+			reg = <0x0 0x1fbf8000 0x0 0x100>;
+
+			clocks = <&i2cclock>;
+
+			/* 100 kHz */
+			clock-frequency = <100000>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			status = "disabled";
+		};
+
+		i2c1: i2c1@1fbf8100 {
+			compatible = "mediatek,mt7621-i2c";
+			reg = <0x0 0x1fbf8100 0x0 0x100>;
+
+			clocks = <&i2cclock>;
+
+			/* 100 kHz */
+			clock-frequency = <100000>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			status = "disabled";
+		};
+
+		snfi: spi@1fa10000 {
 			compatible = "airoha,en7581-snand";
 			reg = <0x0 0x1fa10000 0x0 0x140>,
 			      <0x0 0x1fa11000 0x0 0x160>;
@@ -166,23 +727,27 @@ spi@1fa10000 {
 			spi_nand: nand@0 {
 				compatible = "spi-nand";
 				reg = <0>;
-
 				spi-max-frequency = <50000000>;
 				spi-tx-bus-width = <1>;
 				spi-rx-bus-width = <2>;
 			};
 		};
 
-		scuclk: clock-controller@1fb00000 {
-			compatible = "airoha,en7581-scu";
-			reg = <0x0 0x1fb00000 0x0 0x970>;
-			#clock-cells = <1>;
-			#reset-cells = <1>;
-		};
+		mmc0: mmc@1fa0e000 {
+			compatible = "mediatek,mt7622-mmc";
+			reg = <0x0 0x1fa0e000 0x0 0x1000>,
+			      <0x0 0x1fa0c000 0x0 0x60>;
+			interrupts = <GIC_SPI 170 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&scuclk EN7581_CLK_EMMC>, <&clk25m>;
+			clock-names = "source", "hclk";
+			bus-width = <4>;
+			max-frequency = <52000000>;
+			vmmc-supply = <&vmmc_3v3>;
+			disable-wp;
+			cap-mmc-highspeed;
+			non-removable;
 
-		pbus_csr: syscon@1fbe3400 {
-			compatible = "airoha,en7581-pbus-csr", "syscon";
-			reg = <0x0 0x1fbe3400 0x0 0xff>;
+			status = "disabled";
 		};
 
 		pciephy: phy@1fa5a000 {
@@ -217,10 +782,8 @@ pcie0: pcie@1fc00000 {
 
 			ranges = <0x02000000 0 0x20000000 0x0 0x20000000 0 0x4000000>;
 
-			resets = <&scuclk EN7581_PCIE0_RST>,
-				 <&scuclk EN7581_PCIE1_RST>,
-				 <&scuclk EN7581_PCIE2_RST>;
-			reset-names = "phy-lane0", "phy-lane1", "phy-lane2";
+			resets = <&scuclk EN7581_PCIE0_RST>, <&scuclk EN7581_PCIC_PERSTOUT0_RST>;
+			reset-names = "phy-lane0", "perstout";
 
 			mediatek,pbus-csr = <&pbus_csr 0x0 0x4>;
 
@@ -260,10 +823,8 @@ pcie1: pcie@1fc20000 {
 
 			ranges = <0x02000000 0 0x24000000 0x0 0x24000000 0 0x4000000>;
 
-			resets = <&scuclk EN7581_PCIE0_RST>,
-				 <&scuclk EN7581_PCIE1_RST>,
-				 <&scuclk EN7581_PCIE2_RST>;
-			reset-names = "phy-lane0", "phy-lane1", "phy-lane2";
+			resets = <&scuclk EN7581_PCIE1_RST>, <&scuclk EN7581_PCIC_PERSTOUT1_RST>;
+			reset-names = "phy-lane1", "perstout";
 
 			mediatek,pbus-csr = <&pbus_csr 0x8 0xc>;
 
@@ -285,65 +846,69 @@ pcie_intc1: interrupt-controller {
 			};
 		};
 
-		uart1: serial@1fbf0000 {
-			compatible = "ns16550";
-			reg = <0x0 0x1fbf0000 0x0 0x30>;
-			reg-io-width = <4>;
-			reg-shift = <2>;
-			interrupts = <GIC_SPI 18 IRQ_TYPE_LEVEL_HIGH>;
-			clock-frequency = <1843200>;
-		};
-
-		rng@1faa1000 {
-			compatible = "airoha,en7581-trng";
-			reg = <0x0 0x1faa1000 0x0 0xc04>;
-			interrupts = <GIC_SPI 35 IRQ_TYPE_LEVEL_HIGH>;
-		};
-
-		system-controller@1fbf0200 {
-			compatible = "airoha,en7581-gpio-sysctl", "syscon",
-				     "simple-mfd";
-			reg = <0x0 0x1fbf0200 0x0 0xc0>;
+		pcie2: pcie@1fc40000 {
+			compatible = "airoha,en7581-pcie";
+			device_type = "pci";
+			linux,pci-domain = <2>;
+			#address-cells = <3>;
+			#size-cells = <2>;
 
-			en7581_pinctrl: pinctrl {
-				compatible = "airoha,en7581-pinctrl";
+			reg = <0x0 0x1fc40000 0x0 0x1670>;
+			reg-names = "pcie-mac";
 
-				interrupt-parent = <&gic>;
-				interrupts = <GIC_SPI 26 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&scuclk EN7523_CLK_PCIE>;
+			clock-names = "sys-ck";
 
-				gpio-controller;
-				#gpio-cells = <2>;
+			phys = <&usb1_phy PHY_TYPE_USB3>;
+			phy-names = "pcie-phy";
 
-				interrupt-controller;
-				#interrupt-cells = <2>;
-			};
-		};
+			ranges = <0x02000000 0 0x28000000 0x0 0x28000000 0 0x4000000>;
 
-		i2c0: i2c@1fbf8000 {
-			compatible = "mediatek,mt7621-i2c";
-			reg = <0x0 0x1fbf8000 0x0 0x100>;
+			resets = <&scuclk EN7581_PCIE2_RST>, <&scuclk EN7581_PCIC_PERSTOUT2_RST>;
+			reset-names = "phy-lane2", "perstout";
 
-			resets = <&scuclk EN7581_I2C2_RST>;
+			mediatek,pbus-csr = <&pbus_csr 0x10 0x14>;
 
-			clocks = <&clk20m>;
-			clock-frequency = <100000>;
-			#address-cells = <1>;
-			#size-cells = <0>;
+			interrupts = <GIC_SPI 41 IRQ_TYPE_LEVEL_HIGH>;
+			bus-range = <0x00 0xff>;
+			#interrupt-cells = <1>;
+			interrupt-map-mask = <0 0 0 7>;
+			interrupt-map = <0 0 0 1 &pcie_intc2 0>,
+					<0 0 0 2 &pcie_intc2 1>,
+					<0 0 0 3 &pcie_intc2 2>,
+					<0 0 0 4 &pcie_intc2 3>;
 
 			status = "disabled";
-		};
-
-		i2c1: i2c@1fbf8100 {
-			compatible = "mediatek,mt7621-i2c";
-			reg = <0x0 0x1fbf8100 0x0 0x100>;
-
-			resets = <&scuclk EN7581_I2C_MASTER_RST>;
 
-			clocks = <&clk20m>;
-			clock-frequency = <100000>;
-			#address-cells = <1>;
-			#size-cells = <0>;
+			pcie_intc2: interrupt-controller {
+				interrupt-controller;
+				#address-cells = <0>;
+				#interrupt-cells = <1>;
+			};
+		};
 
+		npu: npu@1e900000 {
+			compatible = "airoha,en7581-npu";
+			reg = <0x0 0x1e900000 0x0 0x313000>;
+			interrupts = <GIC_SPI 125 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 103 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 109 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 113 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 117 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 134 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 135 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 136 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 137 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 118 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 119 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 120 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 121 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 122 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 123 IRQ_TYPE_LEVEL_HIGH>;
+			memory-region = <&npu_binary>, <&npu_pkt>, <&npu_txpkt>,
+					<&npu_txbufid>, <&npu_ba>;
+			memory-region-names = "binary", "pkt", "tx-pkt",
+					      "tx-bufid", "ba";
 			status = "disabled";
 		};
 
@@ -362,9 +927,9 @@ eth: ethernet@1fb50000 {
 				 <&scuclk EN7581_DUAL_HSI1_MAC_RST>,
 				 <&scuclk EN7581_HSI_MAC_RST>,
 				 <&scuclk EN7581_XFP_MAC_RST>;
-			reset-names = "fe", "pdma", "qdma",
-				      "xsi-mac", "hsi0-mac", "hsi1-mac",
-				      "hsi-mac", "xfp-mac";
+			reset-names = "fe", "pdma", "qdma", "xsi-mac",
+				      "hsi0-mac", "hsi1-mac", "hsi-mac",
+				      "xfp-mac";
 
 			interrupts = <GIC_SPI 37 IRQ_TYPE_LEVEL_HIGH>,
 				     <GIC_SPI 55 IRQ_TYPE_LEVEL_HIGH>,
@@ -377,6 +942,11 @@ eth: ethernet@1fb50000 {
 				     <GIC_SPI 49 IRQ_TYPE_LEVEL_HIGH>,
 				     <GIC_SPI 64 IRQ_TYPE_LEVEL_HIGH>;
 
+			memory-region = <&qdma0_buf>, <&qdma1_buf>;
+			memory-region-names = "qdma0-buf", "qdma1-buf";
+
+			airoha,npu = <&npu>;
+
 			status = "disabled";
 
 			#address-cells = <1>;
@@ -394,6 +964,176 @@ fixed-link {
 					pause;
 				};
 			};
+
+			gdm2: ethernet@2 {
+				compatible = "airoha,eth-mac";
+				reg = <2>;
+				pcs-handle = <&pon_pcs>;
+
+				status = "disabled";
+			};
+
+			gdm4: ethernet@4 {
+				compatible = "airoha,eth-mac";
+				reg = <4>;
+				pcs-handle = <&eth_pcs>;
+
+				status = "disabled";
+			};
+		};
+
+		switch: switch@1fb58000 {
+			compatible = "airoha,en7581-switch";
+			reg = <0 0x1fb58000 0 0x8000>;
+			resets = <&scuclk EN7581_GSW_RST>;
+
+			interrupt-controller;
+			#interrupt-cells = <1>;
+			interrupt-parent = <&gic>;
+			interrupts = <GIC_SPI 31 IRQ_TYPE_LEVEL_HIGH>;
+
+			status = "disabled";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			ports {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				gsw_port1: port@1 {
+					reg = <1>;
+					phy-mode = "internal";
+					phy-handle = <&gsw_phy1>;
+					status = "disabled";
+				};
+
+				gsw_port2: port@2 {
+					reg = <2>;
+					phy-mode = "internal";
+					phy-handle = <&gsw_phy2>;
+					status = "disabled";
+				};
+
+				gsw_port3: port@3 {
+					reg = <3>;
+					phy-mode = "internal";
+					phy-handle = <&gsw_phy3>;
+					status = "disabled";
+				};
+
+				gsw_port4: port@4 {
+					reg = <4>;
+					phy-mode = "internal";
+					phy-handle = <&gsw_phy4>;
+					status = "disabled";
+				};
+
+				port@6 {
+					reg = <6>;
+					ethernet = <&gdm1>;
+					phy-mode = "internal";
+
+					fixed-link {
+						speed = <10000>;
+						full-duplex;
+						pause;
+					};
+				};
+			};
+
+			mdio: mdio {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				gsw_phy1: ethernet-phy@9 {
+					compatible = "ethernet-phy-ieee802.3-c22";
+					reg = <9>;
+					phy-mode = "internal";
+					status = "disabled";
+
+					leds {
+						#address-cells = <1>;
+						#size-cells = <0>;
+
+						gsw_phy1_led0: led@0 {
+							reg = <0>;
+							status = "disabled";
+						};
+
+						gsw_phy1_led1: led@1 {
+							reg = <1>;
+							status = "disabled";
+						};
+					};
+				};
+
+				gsw_phy2: ethernet-phy@a {
+					compatible = "ethernet-phy-ieee802.3-c22";
+					reg = <10>;
+					phy-mode = "internal";
+					status = "disabled";
+
+					leds {
+						#address-cells = <1>;
+						#size-cells = <0>;
+
+						gsw_phy2_led0: led@0 {
+							reg = <0>;
+							status = "disabled";
+						};
+
+						gsw_phy2_led1: led@1 {
+							reg = <1>;
+							status = "disabled";
+						};
+					};
+				};
+
+				gsw_phy3: ethernet-phy@b {
+					compatible = "ethernet-phy-ieee802.3-c22";
+					reg = <11>;
+					phy-mode = "internal";
+					status = "disabled";
+
+					leds {
+						#address-cells = <1>;
+						#size-cells = <0>;
+
+						gsw_phy3_led0: led@0 {
+							reg = <0>;
+							status = "disabled";
+						};
+
+						gsw_phy3_led1: led@1 {
+							reg = <1>;
+							status = "disabled";
+						};
+					};
+				};
+
+				gsw_phy4: ethernet-phy@c {
+					compatible = "ethernet-phy-ieee802.3-c22";
+					reg = <12>;
+					phy-mode = "internal";
+					status = "disabled";
+
+					leds {
+						#address-cells = <1>;
+						#size-cells = <0>;
+
+						gsw_phy4_led0: led@0 {
+							reg = <0>;
+							status = "disabled";
+						};
+
+						gsw_phy4_led1: led@1 {
+							reg = <1>;
+							status = "disabled";
+						};
+					};
+				};
+			};
 		};
 	};
 };
diff --git a/arch/mips/boot/dts/econet/en751221.dtsi b/arch/mips/boot/dts/econet/en751221.dtsi
index a413d4887ffe..1d4f10d96dd1 100644
--- a/arch/mips/boot/dts/econet/en751221.dtsi
+++ b/arch/mips/boot/dts/econet/en751221.dtsi
@@ -359,7 +359,7 @@ eth: ethernet@1fb50000 {
 		         <&scuclk EN751221_GDMA_RST>,
 		         <&scuclk EN751221_FE_UNZIP_RST>,
 		         <&scuclk EN751221_GDMP_RST>;
-		reset-names = "fe", "qdma1", "qdma2", /* rename to pdma, qdma */
+		reset-names = "fe", "pdma", "qdma", /* rename to pdma, qdma */
 		              "gdma", "fe-unzip", "gdmp";
 
 		#address-cells = <1>;
diff --git a/arch/mips/boot/dts/econet/en7528.dtsi b/arch/mips/boot/dts/econet/en7528.dtsi
index 499e4e619a43..60c282e2ff45 100644
--- a/arch/mips/boot/dts/econet/en7528.dtsi
+++ b/arch/mips/boot/dts/econet/en7528.dtsi
@@ -492,7 +492,7 @@ eth: ethernet@1fb50000 {
 			         <&scuclk EN7528_FE_QDMA1_RST>,
 			         <&scuclk EN7528_FE_QDMA2_RST>,
 			         <&scuclk EN7528_GDMP_RST>;
-			reset-names = "fe", "qdma0", "qdma1", "gdmp";
+			reset-names = "fe", "pdma", "qdma", "gdmp";
 
 			airoha,wed = <&wed0>, <&wed1>;
 			interrupts = <GIC_SHARED 21 IRQ_TYPE_LEVEL_HIGH>,  /* QDMA LAN0 */
-- 
2.55.0

